src/helpers.rs

```diff
@@ -827,6 +833,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                    }
+                },
+                _ => {
+                    libffi_args.push((Box::new(()), CArg::INVALID));
+                    throw_unsup_format!("Unsupported argument type to external C function: {:?}", cur_arg);
+                }
+            }
+        }
+
+        // convert them to libffi Arg type
+        let libffi_args = libffi_args.iter()
+            .map(|cur_arg| {
+                match cur_arg {
+                    (boxed, CArg::Int8(_)) => {
+                        arg(&(*boxed.downcast_ref::<i8>().unwrap()))
+                    },
+                    (boxed, CArg::Int16(_)) => {
+                        arg(&(*boxed.downcast_ref::<i16>().unwrap()))
+                    },
+                    (boxed, CArg::Int32(_)) => {
+                        arg(&(*boxed.downcast_ref::<i32>().unwrap()))
+                    },
+                    (boxed, CArg::Int64(_)) => {
+                        arg(&(*boxed.downcast_ref::<i64>().unwrap()))
+                    },
+                    (boxed, CArg::UInt8(_)) => {
+                        arg(&(*boxed.downcast_ref::<u8>().unwrap()))
+                    },
+                    (boxed, CArg::UInt16(_)) => {
+                        arg(&(*boxed.downcast_ref::<u16>().unwrap()))
+                    },
+                    (boxed, CArg::UInt32(_)) => {
+                        arg(&(*boxed.downcast_ref::<u32>().unwrap()))
+                    },
+                    (boxed, CArg::UInt64(_)) => {
+                        arg(&(*boxed.downcast_ref::<u64>().unwrap()))
+                    },
+                    _ => {
+                        // should be unreachable
+                        panic!("Trying to call external function with unsupported type: should have already bailed");
+                    }
+                }
+            }).collect::<Vec<libffi::high::Arg<'_>>>();
+
+        // call the C function using libffi
+        unsafe {
+            let lib = match libloading::Library::new(match this.machine.external_c_so_file.as_ref(){
+                Some(x) => x,
+                None => {
+                    throw_unsup_format!("Error reading specified shared object file");
+                }
+            }) {
+                Ok(x) => x,
+                Err(_) => {
+                    throw_unsup_format!("Error reading specified shared object file");
+                }
+            };
+            let func: libloading::Symbol<'_, unsafe extern fn()> = match lib.get(link_name.as_str().as_bytes()) {
+                Ok(x) => x,
+                Err(_) => {
+                    throw_unsup_format!("Error reading function {:?} from specified shared object file", link_name.as_str());
+                }
+            };
+            // code pointer to C function
+            let ptr = CodePtr(*func.deref() as *mut _);
+            // store output, depending on return type in the function signature
+            match external_fct_defn.output_type {
```

Comment:
Factor out into a helper, consider macros to reduce duplication.

Reply:
good point